x/programs/rust/sdk-macros/src/lib.rs

```diff
@@ -179,74 +195,75 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
+                Self::#variant_ident(a) => wasmlanche_sdk::state::PrefixedBytes::new(#idx, a.as_ref())
+            },
+            // ex: Point
+            Fields::Unit => quote! {
+                Self::#variant_ident => wasmlanche_sdk::state::PrefixedBytes::new(#idx, &[])
+            },
+            // ex: Point { x: f64, y: f64 }
+            Fields::Named(_) => quote! {
+                Self::#variant_ident { .. } => panic!("named enum fields are not supported")
+            },
         }
+    });
 
-        // Generate the Into<key> implementation needed to
-        // convert the enum to a Key type.
-        impl Into<wasmlanche_sdk::state::Key> for #name {
-            fn into(self) -> wasmlanche_sdk::state::Key {
-                wasmlanche_sdk::state::Key::new(self.to_vec())
-            }
-        }
+    let trait_implementation_body = if !variants.is_empty() {
+        quote! { match self { #(#match_arms),* } }
+    } else {
+        quote! { wasmlanche_sdk::state::PrefixedBytes::new(0, &[]) }
     };
 
-    TokenStream::from(gen)
-}
+    quote! {
+        #item_enum
 
-fn generate_to_vec(
-    variants: &syn::punctuated::Punctuated<syn::Variant, syn::Token![,]>,
-) -> Vec<proc_macro2::TokenStream> {
-    variants
-        .iter()
-        .enumerate()
-        .map(|(idx, variant)| {
-            let variant_ident = &variant.ident;
-            let index = idx as u8;
-            match &variant.fields {
-                // ex: Point(f64, f64)
-                Fields::Unnamed(_) => quote! {
-                    Self::#variant_ident(a) => std::iter::once(#index).chain(a.into_iter()).collect()
-                },
-                // ex: Point
-                Fields::Unit => quote! {
-                    Self::#variant_ident => vec![#index]
-                },
-                // ex: Point { x: f64, y: f64 }
-                Fields::Named(_) => quote! {
-                    Self::#variant_ident { .. } => panic!("named enum fields are not supported"),
-                },
+        impl wasmlanche_sdk::state::Key for #name {
+            fn as_prefixed(&self) -> wasmlanche_sdk::state::PrefixedBytes<'_> {
+                #trait_implementation_body
             }
-        })
-        .collect()
+        }
+    }
+    .into()
 }
 
 /// Returns whether the type_path represents a Program type.
 fn is_context(type_path: &std::boxed::Box<Type>) -> bool {
     if let Type::Path(type_path) = type_path.as_ref() {
-        type_path.path.segments.last() == parse_str::<Path>(CONTEXT_TYPE).unwrap().segments.last()
+        let mut context_type = parse_str::<Path>(CONTEXT_TYPE).unwrap();
+
+        if let Some(segment) = type_path.path.segments.last() {
+            let context_type = context_type.segments.last_mut();
+            context_type.unwrap().arguments = segment.arguments.clone();
+        }
+
+        type_path.path.segments.last() == context_type.segments.last()
     } else {
         false
     }
```

Comment:
I would approve a PR with this haha. I don't quite think it's necessary right now because the `ui` tests cover the cases pretty well, but more unit tests usually make things easier to expand and extend.